pkg/controllers/jobset_controller_test.go

```diff
@@ -101,3 +105,163 @@ func TestIsJobFinished(t *testing.T) {
 		})
 	}
 }
+
+func TestSetExclusiveAffinities(t *testing.T) {
+	var (
+		topologyKey = "test-topology-key"
+		jobName     = "test-job"
+	)
+	tests := []struct {
+		name         string
+		job          *batchv1.Job
+		nsSelector   *metav1.LabelSelector
+		wantAffinity corev1.Affinity
+	}{
+		{
+			name:       "no existing affinities",
+			job:        testutils.MakeJob(jobName).Obj(),
+			nsSelector: &metav1.LabelSelector{}, // all namespaces
+			wantAffinity: corev1.Affinity{
+				PodAffinity: &corev1.PodAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+								{
+									Key:      jobset.JobNameKey,
+									Operator: metav1.LabelSelectorOpIn,
+									Values:   []string{jobName},
+								},
+							}},
+							TopologyKey:       topologyKey,
+							NamespaceSelector: &metav1.LabelSelector{},
+						},
+					},
+				},
+				PodAntiAffinity: &corev1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+								{
+									Key:      jobset.JobNameKey,
+									Operator: metav1.LabelSelectorOpExists,
+								},
+								{
+									Key:      jobset.JobNameKey,
+									Operator: metav1.LabelSelectorOpNotIn,
+									Values:   []string{jobName},
+								},
+							}},
+							TopologyKey:       topologyKey,
+							NamespaceSelector: &metav1.LabelSelector{},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "existing affinities should be appended to, not replaced",
+			job: testutils.MakeJob(jobName).SetAffinity(&corev1.Affinity{
+				PodAffinity: &corev1.PodAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+								{
+									Key:      "label-foo",
+									Operator: metav1.LabelSelectorOpIn,
+									Values:   []string{"value-foo"},
+								},
+							}},
+							TopologyKey: "topology.kubernetes.io/zone",
+						},
+					},
+				},
+				PodAntiAffinity: &corev1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+						{
+							LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+								{
+									Key:      "label-bar",
+									Operator: metav1.LabelSelectorOpNotIn,
+									Values:   []string{"value-bar"},
+								},
+							}},
+							TopologyKey: "topology.kubernetes.io/zone",
+						},
+					},
+				},
+			}).Obj(),
+			nsSelector: &metav1.LabelSelector{},
```

Comment:
just to use something different
```suggestion
			nsSelector: nil
```